DMCompiler/DM/Builders/DMExpressionBuilder.cs

```diff
@@ -644,11 +644,16 @@ private DMExpression BuildScopeIdentifier(DMASTScopeIdentifier scopeIdentifier,
         if (scopeIdentifier.Expression is DMASTIdentifier { Identifier: "type" or "parent_type" } identifier) {
             // This is the same behaviour as in BYOND, but BYOND simply raises an undefined var error.
             // We want to give end users an explanation at least.
-            if (scopeMode is Normal && ctx.Proc != null)
-                return BadExpression(WarningCode.BadExpression, identifier.Location,
-                    "Use of \"type::\" and \"parent_type::\" outside of a context is forbidden");
-
-            if (identifier.Identifier == "parent_type") {
+            if (scopeMode is Normal && ctx.Proc != null) {
+                if (ctx.Proc.GetLocalVariable(identifier.Identifier) != null) {
+                    // actually - it's referring to a local variable named "type" or "parent_type"... just do the usual thing
+                    expression = BuildExpression(scopeIdentifier.Expression, inferredPath);
```

Comment:
We should emit a warning for this, since it's likely unintentional.

Reply:
I did `SoftResolvedKeyword` which made goon [fail to compile](https://github.com/OpenDreamProject/OpenDream/actions/runs/13489680448/job/37685809461?pr=2227#step:16:19) due to its `#pragma SoftReservedKeyword error`. Let me know if I should pick something else.

Reply:
After discussing, went with a new and specific `OD2001: ScopeOperandNamedType`